### PR TITLE
ENH: working threaded implementation

### PR DESCRIPTION
**Pull request by tacaswell**
can register and search (so far)

### CODE REVIEW

File: caproto/thread/client.py [hunk 1]
```diff
@@ -0,0 +1,681 @@
...
+
+        self.has_new_command = threading.Condition()
+        self.sock_thread = SocketThread(sock, self)
+
+    def send(self, port, *commands):
+        """
+        Process a command and tranport it over the UDP socket.
+        """
+        bytes_to_send = self.broadcaster.send(*commands)
+        for host in ca.get_address_list() + ['0.0.0.0']:
```
Comment (danielballan) [inline caproto/thread/client.py]: Note so we dont' forget to delete this

File: caproto/thread/client.py [hunk 2]
```diff
@@ -0,0 +1,681 @@
...
+            print('sending to', (host, port), bytes_to_send)
+            self.udp_sock.sendto(bytes_to_send, (host, port))
+
+    def recv(self):
```
Comment (danielballan) [inline caproto/thread/client.py]: Remove.

File: caproto/thread/client.py [hunk 3]
```diff
@@ -0,0 +1,681 @@
...
+        ver_command, search_command = self.broadcaster.search(name)
+        # Stash the search ID for recognizes the SearchResponse later.
+        self.unanswered_searches[search_command.cid] = name
+        self.send(ca.EPICS_CA1_PORT, ver_command, search_command)
+        # Wait for the SearchResponse.
+        while True:
+            with self.has_new_command:
+                if search_command.cid not in self.unanswered_searches:
+                    break
+                suc = self.has_new_command.wait(5)
```
Comment (danielballan) [inline caproto/thread/client.py]: Make this magic number a configurable attribute.

File: caproto/thread/client.py [hunk 4]
```diff
@@ -0,0 +1,681 @@
...
+                self.has_new_command.notify_all()
+
+
+class VirtualCircuit:
+    def __init__(self, circuit):
+        self.circuit = circuit  # a caproto.VirtualCircuit
+        self.channels = {}  # map cid to Channel
+        self.ioids = {}  # map ioid to Channel
+        self.subscriptionids = {}  # map subscriptionid to Channel
+        self.event = None  # used for signaling consumers about new commands
```
Comment (danielballan) [inline caproto/thread/client.py]: Remove this.

**Comment by tacaswell** [comment]
This passes locally, not sure why this is failing here...

**Pull request merged**